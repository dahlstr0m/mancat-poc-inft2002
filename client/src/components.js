// @flow

import * as React from 'react';
import { Component } from 'react-simplified';
import { NavLink } from 'react-router-dom';
import { createHashHistory } from 'history';
import {
  Alert,
  Card,
  Row,
  Column,
  Form,
  Button,
  NavBar,
  PortfolioCard,
  ProjectCard,
  PosterCard,
  CardImage,
  CardColumn,
  CardGrid,
  CardPlain,
  CardBody,
} from './widgets';
import {
  projectService,
  categoryService,
  posterService,
  employerService,
  type Project,
  type Category,
  type Poster,
  type Employer,
} from './portfolio-service';

const history = createHashHistory(); // Use history.push(...) to programmatically change path, for instance after successfully saving a student

/**
 * Renders portfolio list.
 */
export class PortfolioListing extends Component {
  projects: Project[] = [];
  posters: Poster[] = [];

  render() {
    return (
      <>
        {this.projects
          .filter((project) => project.active === true) // Filter to ensure only Active projects to be displayed
          .sort((projectA, projectB) => projectA.ranking - projectB.ranking) // Sorting to display by asc ranking
          .map((
            project // Mapping to display all left projects with PortfolioCard widget & fetching correct corresponding thumbnail for url attrib.
          ) => (
            <PortfolioCard
              projectId={project.projectId}
              title={project.title}
              link={'/projects/' + project.projectId}
              imageUrl={
                this.posters.find((poster) => poster.projectId === project.projectId)
                  ? this.posters.find((poster) => poster.projectId === project.projectId)
                      .thumbnailUrl
                  : ''
              }
            >
              {'ingen data'}
            </PortfolioCard>
          ))}
      </>
    );
  }

  mounted() {
    projectService
      .getProjects()
      .then((projects) => (this.projects = projects))
      .catch((error: Error) => Alert.danger('Error getting projects: ' + error.message));

    posterService
      .getPosters()
      .then((posters) => (this.posters = posters))
      .catch((error: Error) => Alert.danger('Error getting posters: ' + error.message));
  }
}

/**
 * Renders project details.
 */
export class ProjectDetails extends Component<{ match: { params: { id: number } } }> {
  project: Project = {
    projectId: 0,
    title: '',
    description: '',
    date: '',
    categoryId: 0,
    employerId: 0,
    active: false,
    ranking: 0,
  };
  category: Category = { categoryId: 0, name: '' };
  employer: Employer = { employerId: 0, name: '' };
  posters: Poster[] = [];

  render() {
    return (
      <>
        <ProjectCard
          projectId={this.project.projectId}
          title={this.project.title}
          description={this.project.description}
          date={this.project.date}
          category={this.category.name}
          employer={this.employer.name}
        >
          {this.posters.map((poster) => (
            <PosterCard
              posterId={poster.posterId}
              description={poster.description}
              url={poster.url}
              thumbnailUrl={poster.thumbnailUrl}
            >
              {'ingen data'}
            </PosterCard>
          ))}
        </ProjectCard>
      </>
    );
  }
  mounted() {
    projectService
      .getProject(this.props.match.params.id)
      .then((project) => (this.project = project))
      .catch((error: Error) => Alert.danger('Error getting project: ' + error.message));

    categoryService
      .getProjectCategory(this.props.match.params.id)
      .then((category) => (this.category = category))
      .catch((error: Error) => Alert.danger('Error getting category: ' + error.message));

    employerService
      .getProjectEmployer(this.props.match.params.id)
      .then((employer) => (this.employer = employer))
      .catch((error: Error) => Alert.danger('Error getting employer: ' + error.message));

    posterService
      .getProjectPosters(this.props.match.params.id)
      .then((posters) => (this.posters = posters))
      .catch((error: Error) => Alert.danger('Error getting posters: ' + error.message));
  }
}

export class OtherManagement extends Component {
  employers: Employer[] = [];
  categories: Category[] = [];

  render() {
    return (
      <CardPlain>
        <CardBody title="Manage categories">
          <Button.Success>Manage</Button.Success>
        </CardBody>
        <hr />
        <CardBody title="Manage employers">
          <Button.Success>Manage</Button.Success>
        </CardBody>
        <hr />
        <CardBody title="Add new project">
          <Button.Success>Manage</Button.Success>
        </CardBody>
      </CardPlain>
    );
  }
}

export class ProjectManagement extends Component {
  projects: Project[] = [];
  posters: Poster[] = [];

  render() {
    return (
      <>
        <h1>Project management</h1>
        <CardGrid columns={1} columnsSm={2} columnsMd={4}>
          {this.projects.map((project) => (
            <CardColumn>
              <CardImage
                img={
                  this.posters.find((poster) => poster.projectId === project.projectId)
                    ? this.posters.find((poster) => poster.projectId === project.projectId)
                        .thumbnailUrl
                    : ''
                }
                title={project.title}
                alt={project.title}
                buttonText={'Manage project'}
                buttonOnClick={() => history.push(`/admin/projects/${project.projectId}`)}
              >
                {project.description}
              </CardImage>
            </CardColumn>
          ))}
        </CardGrid>
      </>
    );
  }

  mounted() {
    projectService
      .getProjects()
      .then((projects) => (this.projects = projects))
      .catch((error: Error) => Alert.danger('Error getting projects: ' + error));

    posterService
      .getPosters()
      .then((posters) => (this.posters = posters))
      .catch((error: Error) => Alert.danger('Error getting posters: ' + error));
  }
}
