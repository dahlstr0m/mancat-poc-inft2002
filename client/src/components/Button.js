// @flow

import * as React from 'react';
import { Component } from 'react-simplified';

/**
 * Renders a success button using Bootstrap styles.
 */
class ButtonSuccess extends Component<{
  onClick: () => mixed,
  small?: boolean,
  children?: React.Node,
}> {
  render() {
    return (
      <button
        type="button"
        className={'btn btn-success' + (this.props.small ? ' btn-sm py-0' : '')}
        onClick={this.props.onClick}
      >
        {this.props.children}
      </button>
    );
  }
}

/**
 * Renders a danger button using Bootstrap styles.
 */
class ButtonDanger extends Component<{
  onClick: () => mixed,
  small?: boolean,
  children?: React.Node,
}> {
  render() {
    return (
      <button
        type="button"
        className={'btn btn-danger' + (this.props.small ? ' btn-sm py-0' : '')}
        onClick={this.props.onClick}
      >
        {this.props.children}
      </button>
    );
  }
}

/**
 * Renders a light button using Bootstrap styles.
 */
class ButtonLight extends Component<{
  onClick: () => mixed,
  small?: boolean,
  children?: React.Node,
}> {
  render() {
    return (
      <button
        type="button"
        className={'btn btn-light' + (this.props.small ? ' btn-sm py-0' : '')}
        onClick={this.props.onClick}
      >
        {this.props.children}
      </button>
    );
  }
}

/**
 * Renders a button using Bootstrap styles.
 */
export default class Button {
  static Success = ButtonSuccess;
  static Danger = ButtonDanger;
  static Light = ButtonLight;
}
