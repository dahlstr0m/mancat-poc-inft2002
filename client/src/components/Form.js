// @flow

import * as React from 'react';
import { Component } from 'react-simplified';

/**
 * Renders a form label using Bootstrap styles.
 */
class FormLabel extends Component<{ children?: React.Node }> {
  render() {
    return <label className="col-form-label">{this.props.children}</label>;
  }
}

/**
 * Renders a form input using Bootstrap styles.
 */
class FormInput extends Component<{
  type: string,
  value: React.Node,
  onChange: (SyntheticEvent<HTMLInputElement>) => mixed,
}> {
  render() {
    // ...rest will contain extra passed attributes such as disabled, required, width, height, pattern
    // For further information, see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax
    const { type, value, onChange, ...rest } = this.props;
    return (
      <input
        {...rest}
        className="form-control"
        type={this.props.type}
        value={this.props.value}
        onChange={this.props.onChange}
      />
    );
  }
}

/**
 * Renders a form textarea using Bootstrap styles.
 */
class FormTextarea extends React.Component<{
  value: React.Node,
  onChange: (SyntheticEvent<HTMLTextAreaElement>) => mixed,
}> {
  render() {
    // ...rest will contain extra passed attributes such as disabled, required, rows, cols
    // For further information, see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax
    const { value, onChange, ...rest } = this.props;
    return <textarea {...rest} className="form-control" value={value} onChange={onChange} />;
  }
}

/**
 * Renders a form checkbox using Bootstrap styles.
 */
class FormCheckbox extends Component<{
  checked: React.Node,
  onChange: (SyntheticEvent<HTMLInputElement>) => mixed,
  labeling: React.Node,
}> {
  render() {
    // ...rest will contain extra passed attributes such as disabled, required, width, height, pattern
    // For further information, see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax
    const { labeling, checked, onChange, ...rest } = this.props;
    return (
      <div className="custom-control custom-checkbox">
        <input
          {...rest}
          type="checkbox"
          className="custom-control-input"
          id="customCheck"
          checked={checked}
          onChange={onChange}
        />
        <label className="custom-control-label" htmlFor="customCheck">
          {labeling}
        </label>
      </div>
    );
  }
}

/**
 * Renders a form select using Bootstrap styles.
 */
class FormSelect extends Component<{
  value: React.Node,
  onChange: (SyntheticEvent<HTMLSelectElement>) => mixed,
  children?: React.Node,
}> {
  render() {
    // ...rest will contain extra passed attributes such as disabled, required, size.
    // For further information, see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax
    const { value, onChange, children, ...rest } = this.props;
    return (
      <select {...rest} className="custom-select" value={value} onChange={onChange}>
        {children}
      </select>
    );
  }
}

/**
 * Renders a form date using Bootstrap styles.
 */
class FormDate extends Component<{
  type: string,
  value: React.Node,
  onChange: (SyntheticEvent<HTMLInputElement>) => mixed,
}> {
  render() {
    // ...rest will contain extra passed attributes such as disabled, required, width, height, pattern
    // For further information, see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax
    const { type, value, onChange, ...rest } = this.props;
    return (
      <input
        {...rest}
        className="form-control"
        type={this.props.type}
        value={this.props.value}
        onChange={this.props.onChange}
      />
    );
  }
}

/**
 * Renders form components using Bootstrap styles.
 */
export default class Form {
  static Label = FormLabel;
  static Input = FormInput;
  static Textarea = FormTextarea;
  static Checkbox = FormCheckbox;
  static Select = FormSelect;
  static Date = FormDate;
}
