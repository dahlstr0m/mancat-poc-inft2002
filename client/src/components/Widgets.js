// @flow

import * as React from 'react';
import { Component } from 'react-simplified';
import { NavLink } from 'react-router-dom';

/**
 * Renders alert messages using Bootstrap classes.
 */
export class Alert extends Component {
  alerts: { id: number, text: React.Node, type: string }[] = [];
  nextId = 0;

  render() {
    return (
      <>
        {this.alerts.map((alert, i) => (
          <div key={alert.id} className={'alert alert-' + alert.type} role="alert">
            {alert.text}
            <button type="button" className="close" onClick={() => this.alerts.splice(i, 1)}>
              &times;
            </button>
          </div>
        ))}
      </>
    );
  }

  /**
   * Show success alert.
   */
  static success(text: React.Node) {
    // To avoid 'Cannot update during an existing state transition' errors, run after current event through setTimeout
    setTimeout(() => {
      const instance = Alert.instance(); // Get rendered Alert component instance
      instance?.alerts.push({ id: instance.nextId++, text: text, type: 'success' });
    });
  }

  /**
   * Show info alert.
   */
  static info(text: React.Node) {
    // To avoid 'Cannot update during an existing state transition' errors, run after current event through setTimeout
    setTimeout(() => {
      const instance = Alert.instance(); // Get rendered Alert component instance
      instance?.alerts.push({ id: instance.nextId++, text: text, type: 'info' });
    });
  }

  /**
   * Show warning alert.
   */
  static warning(text: React.Node) {
    // To avoid 'Cannot update during an existing state transition' errors, run after current event through setTimeout
    setTimeout(() => {
      const instance = Alert.instance(); // Get rendered Alert component instance
      instance?.alerts.push({ id: instance.nextId++, text: text, type: 'warning' });
    });
  }

  /**
   * Show danger alert.
   */
  static danger(text: React.Node) {
    // To avoid 'Cannot update during an existing state transition' errors, run after current event through setTimeout
    setTimeout(() => {
      const instance = Alert.instance(); // Get rendered Alert component instance
      instance?.alerts.push({ id: instance.nextId++, text: text, type: 'danger' });
    });
  }
}

/**
 * Renders a row using Bootstrap classes.
 */
export class Row extends Component<{ children?: React.Node }> {
  render() {
    return <div className="row">{this.props.children}</div>;
  }
}

/**
 * Renders a column with specified width using Bootstrap classes.
 */
export class Column extends Component<{ width?: number, right?: boolean, children?: React.Node }> {
  render() {
    return (
      <div
        className={
          'col' +
          (this.props.width ? '-' + this.props.width : '') +
          (this.props.right ? ' text-right' : '')
        }
      >
        {this.props.children}
      </div>
    );
  }
}

/**
 * Renders a NavBar link using Bootstrap styles.
 */
class NavBarLink extends Component<{ to: string, children?: React.Node }> {
  render() {
    return (
      <NavLink className="nav-link" activeClassName="active" to={this.props.to}>
        {this.props.children}
      </NavLink>
    );
  }
}

/**
 * Renders a NavBar using Bootstrap classes.
 */
export class NavBar extends Component<{ brand?: React.Node, children?: React.Node }> {
  static Link = NavBarLink;

  render() {
    return (
      <nav className="navbar navbar-expand-sm bg-light navbar-light">
        {
          <NavLink className="navbar-brand" activeClassName="active" exact to="/">
            {this.props.brand}
          </NavLink>
        }
        <ul className="navbar-nav">{this.props.children}</ul>
      </nav>
    );
  }
}
