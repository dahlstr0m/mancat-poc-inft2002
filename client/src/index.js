// @flow

import ReactDOM from 'react-dom';
import * as React from 'react';
import { Component } from 'react-simplified';
import { HashRouter, Route } from 'react-router-dom';
import { NavBar, Card, Alert, ContactCard, FooterCard } from './widgets';
import { Portfoliolisting, ProjectDetails } from './components';

class Topbar extends Component {
  render() {
    return (
      <NavBar brand="Mancat">
        <NavBar.Link to="/">Sorter</NavBar.Link>
        <NavBar.Link to="/">Portfolio</NavBar.Link>
        <NavBar.Link to="/contact">Contact</NavBar.Link>
      </NavBar>
    );
  }
}

class Contact extends Component {
  render() {
    return <ContactCard title="Contact">This is contactinfo</ContactCard>;
  }
}

class Footer extends Component {
  render() {
    return (
      <FooterCard title="Footer">Laget av Kevin, Hans Petter, Henrik, Bjarne og Mathias</FooterCard>
    );
  }
}

const root = document.getElementById('root');
if (root)
  ReactDOM.render(
    <HashRouter>
      <div>
        <Route // Main page
          exact
          path="/"
          component={() => (
            <div>
              <Topbar />
              <Alert />
              <Portfoliolisting />
              <Footer />
            </div>
          )}
        />
        <Route // Contact page
          exact
          path="/contact"
          component={() => (
            <div>
              <Topbar />
              <Alert />
              <Contact />
              <Footer />
            </div>
          )}
        />
        <Route //Project page
          exact
          path="/project/:id(\d+)"
          component={(props) => (
            <div>
              <Topbar />
              <Alert />
              <ProjectDetails pathid={props.match.params.id} />
              <Portfoliolisting />
              <Footer />
            </div>
          )}
        />
      </div>
    </HashRouter>,
    root
  );
