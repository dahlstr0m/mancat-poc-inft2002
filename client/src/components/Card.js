// @flow

import * as React from 'react';
import { Component } from 'react-simplified';
import { NavLink } from 'react-router-dom';
import Button from './Button';

/**
 * Renders an information card using Bootstrap classes.
 */
export class Card extends Component<{ title?: React.Node, children?: React.Node }> {
  render() {
    return (
      <div className="card">
        <div className="card-body">
          <h5 className="card-title">{this.props.title}</h5>
          <div className="card-text">{this.props.children}</div>
        </div>
      </div>
    );
  }
}

export class CardPlain extends Component<{ children?: React.Node }> {
  render() {
    return (
      <div className="card">
        <div className="card-body">
          <div className="card-text">{this.props.children}</div>
        </div>
      </div>
    );
  }
}

export class TitleCard extends Component<{ img?: React.Node, children?: React.Node }> {
  render() {
    return (
      <div className="card">
        <h1>
          <img
            src={this.props.img}
            alt={'-No img-'}
            width={140}
            height={140}
            style={{ marginRight: '15px' }}
          />
          {this.props.children}
        </h1>
      </div>
    );
  }
}

export class CardBody extends Component<{ title?: React.Node, children?: React.Node }> {
  render() {
    return (
      <div className="card-body">
        <h5 className="card-title">{this.props.title}</h5>
        <div className="card-text">{this.props.children}</div>
      </div>
    );
  }
}

export class CardImage extends Component<{
  img?: React.Node,
  imgAlt?: React.Node,
  imgWidth?: number,
  imgHeight?: number,
  title?: React.Node,
  buttonText?: React.Node,
  buttonOnClick: () => mixed,
  children?: React.Node,
}> {
  render() {
    return (
      <div className="card">
        <div className="card-body">
          <img
            src={this.props.img}
            className="rounded mx-auto d-block"
            alt={this.props.imgAlt}
            width={this.props.imgWidth}
            height={this.props.imgHeight}
          />
          <hr />
          <h5 className="card-title mx-auto d-block">{this.props.title}</h5>
          <div className="card-text">{this.props.children}</div>
          <Button.Danger onClick={this.props.buttonOnClick}>{this.props.buttonText}</Button.Danger>
        </div>
      </div>
    );
  }
}

export class CardImageTwo extends Component<{
  img?: React.Node,
  imgAlt?: React.Node,
  imgWidth?: number,
  imgHeight?: number,
  title?: React.Node,
  children?: React.Node,
  link?: string,
}> {
  render() {
    return (
      <div className="card hover-shadow">
        <div className="card-body">
        <NavLink to={this.props.link}>
          <img
            src={this.props.img}
            className="rounded mx-auto d-block hover-shadow"
            alt={this.props.imgAlt}
            width={this.props.imgWidth}
            height={this.props.imgHeight}
          />
        </NavLink>
        </div>
      </div>
    );
  }
}

export class CardColumn extends Component<{
  children?: React.Node,
}> {
  render() {
    return <div className="col mb-4">{this.props.children}</div>;
  }
}

export class CardGrid extends Component<{
  columns: number,
  columnsSm?: number,
  columnsMd?: number,
  columnsLg?: number,
  children?: React.Node,
}> {
  render() {
    return (
      <div
        className={
          'row row-cols-' +
          this.props.columns.toString() +
          (this.props.columnsSm ? ' row-cols-sm-' + this.props.columnsSm.toString() : '') +
          (this.props.columnsMd ? ' row-cols-md-' + this.props.columnsMd.toString() : '') +
          (this.props.columnsLg ? ' row-cols-lg-' + this.props.columnsLg.toString() : '')
        }
      >
        {this.props.children}
      </div>
    );
  }
}

/**
 * Renders a PortfolioCard for portfoliolisting
 */
export class PortfolioCard extends Component<{
  projectId: number,
  title: string,
  link: string,
  imageUrl?: string,
  children?: React.Node,
}> {
  render() {
    const displaystyleDev = {
      width: '200px',
      backgroundColor: 'lightgrey',
      margin: '20px',
      border: '1px',
    };
    return (
      <>
        <div id={'PortfolioCard' + this.props.projectId} style={displaystyleDev}>
          <NavLink to={this.props.link}>
            <img src={this.props.imageUrl} alt={'Thumbnail kan ikke vises for ' + this.props.title} />
          </NavLink>
        </div>
      </>
    );
  }
}

/**
 * Renders a ProjectCard for ProjectDetails
 */

export class ProjectCard extends Component<{
  projectId?: React.Node,
  title?: React.Node,
  description?: React.Node,
  date?: React.Node,
  category?: React.Node,
  employer?: React.Node,
  children?: React.Node,
}> {
  render() {
    const displaystyleDev2 = {
      width: '100%',
      height: '80%',
      backgroundColor: 'white',
      margin: '20px',
    };
    return (
      <>
        <div style={displaystyleDev2}>
          <p>Tittel: {this.props.title}</p>
          <p>Beskrivelse: {this.props.description}</p>
          <p>Dato: {this.props.date}</p>
          <p>Kategori: {this.props.category}</p>
          <p>Employer: {this.props.employer}</p>
          <div>{this.props.children}</div>
        </div>
      </>
    );
  }
}

/**
 * Renders a PosterCard for ProjectCard
 */

export class PosterCard extends Component<{
  posterId?: React.Node,
  description?: React.Node,
  url?: React.Node,
  thumbnailUrl?: React.Node,
}> {
  render() {
    return (
      <>
        <div id={this.props.posterId}>
          <p>Poster: {this.props.posterId}</p>
          <img src={this.props.url} alt={''} />
          <p>Beskrivelse: {this.props.description}</p>
          <p>Url: {this.props.url}</p>
          <p>ThumbnailUrl: {this.props.thumbnailUrl}</p>
          <hr />
        </div>
      </>
    );
  }
}

/**
 * Renders a ContactCard
 */
export class ContactCard extends Component<{
  title?: React.Node,
  children?: React.Node,
}> {
  render() {
    const displaystyleDev = {
      width: '100%',
      height: '250px',
      backgroundColor: 'lightgreen',
      margin: '20px',
    };
    return (
      <>
        <div id={'Contact'} style={displaystyleDev}>
          <h1>{this.props.title}</h1>
          <div>{this.props.children}</div>
        </div>
      </>
    );
  }
}

/**
 * Renders a Footer
 */
export class FooterCard extends Component<{
  title?: React.Node,
  children?: React.Node,
}> {
  render() {
    const displaystyleDev4 = {
      width: '100%',
      margin: '20px',
    };
    return (
      <>
        <div id={'Footer'} style={displaystyleDev4}>
          <h1>{this.props.title}</h1>
          <div>{this.props.children}</div>
        </div>
      </>
    );
  }
}
