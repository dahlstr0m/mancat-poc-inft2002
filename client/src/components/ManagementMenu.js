// @flow

import * as React from 'react';
import { Component } from 'react-simplified';
import { history } from '../index';
import Button from './Button';
import { CardGrid, CardColumn, CardBody, Card } from './Card';
import { Alert } from './Widgets';

/**
 * Renders All management menu .
 */
export default class ManagementMenu extends Component {
  render() {
    return (
      <CardGrid columns={1} columnsSm={2} columnsMd={3} columnsLg={4}>
        <CardColumn>
          <CardBody title="Projects">
            <Card title="Add new">
              <Button.Success onClick={() => history.push('/admin/projects/new')}>
                Create
              </Button.Success>
            </Card>
          </CardBody>
          <CardBody>
            <Card title="Rearrange order">
              <Button.Light
                onClick={() => Alert.info('This feature is not yet fully implemented. ')}
              >
                Manage
              </Button.Light>
            </Card>
          </CardBody>
        </CardColumn>
        <CardColumn>
          <CardBody title="Poster">
            <Card title="Add new poster">
              <Button.Success onClick={() => history.push('/admin/posters/new')}>
                Create
              </Button.Success>
            </Card>
          </CardBody>
          <CardBody>
            <Card title="Existing posters">
              <Button.Danger onClick={() => history.push('/admin/posters/manage')}>
                Edit
              </Button.Danger>
            </Card>
          </CardBody>
        </CardColumn>
        <CardColumn>
          <CardBody title="Employers">
            <Card title="Manage employers">
              <Button.Danger onClick={() => history.push('/admin/employers')}>Edit</Button.Danger>
            </Card>
          </CardBody>
        </CardColumn>
        <CardColumn>
          <CardBody title="Categories">
            <Card title="Manage Categories">
              <Button.Danger onClick={() => history.push('/admin/categories')}>Edit</Button.Danger>
            </Card>
          </CardBody>
        </CardColumn>
      </CardGrid>
    );
  }
}
