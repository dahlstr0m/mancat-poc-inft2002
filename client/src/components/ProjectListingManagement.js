// @flow

import * as React from 'react';
import { Component } from 'react-simplified';
import { history } from '../index';
import { CardGrid, CardColumn, CardImage } from './Card';
import { Alert } from './Widgets';
import {
  projectService,
  posterService,
  type Project,
  type Poster,
} from '../services/portfolio-service';

/**
 * Renders project listing in management
 */
export default class ProjectListingManagement extends Component {
  projects: Project[] = [];
  posters: Poster[] = [];

  render() {
    return (
      <>
        <h1>Project management</h1>
        <CardGrid columns={1} columnsSm={2} columnsMd={3} columnsLg={4}>
          {this.projects
            .sort((projectA, projectB) => projectA.ranking - projectB.ranking) // Sorting to display by asc ranking
            .map((project) => (
              <CardColumn key={project.projectId}>
                <CardImage
                  button
                  img={this.getPosterUrl(project)}
                  title={project.title}
                  imgAlt={'Missing thumbnail for ' + project.title}
                  buttonText={'Manage project'}
                  buttonOnClick={() => history.push(`/admin/projects/${project.projectId}`)}
                >
                  Description: {project.projectDescription}
                  <br />
                  Status: {project.active == true ? 'Active' : 'Disabled'}
                </CardImage>
              </CardColumn>
            ))}
        </CardGrid>
      </>
    );
  }

  getPosterUrl(project: Project) {
    const poster = this.posters.find((poster) => poster.projectId === project.projectId);
    return poster ? poster.thumbnailUrl : '';
  }

  mounted() {
    projectService
      .getProjects()
      .then((projects) => (this.projects = projects))
      .catch((error: Error) => Alert.danger('Error getting projects: ' + error.message));

    posterService
      .getPosters()
      .then((posters) => (this.posters = posters))
      .catch((error: Error) => Alert.danger('Error getting posters: ' + error.message));
  }
}
