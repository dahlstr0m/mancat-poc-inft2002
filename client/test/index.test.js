// @flow

import * as React from 'react';
import { shallow } from 'enzyme';

// Components
import { Topbar, BackToAdmin, Contact, Footer, AdminPage } from '../src/index';
import { NavBar } from '../src/components/Widgets';
import Button from '../src/components/Button';
import { Card, TitleCard } from '../src/components/Card';
import ManagementMenu from '../src/components/ManagementMenu';
import ProjectListingManagement from '../src/components/ProjectListingManagement';

// Tests
describe('Main components tests', () => {
  test('Topbar component draws correctly', () => {
    const wrapper = shallow(<Topbar />);

    expect(
      wrapper.contains(
        <NavBar brand="Mancat">
          <NavBar.Link to="/">Portfolio</NavBar.Link>
          <NavBar.Link to="/contact">Contact</NavBar.Link>
          <NavBar.Link to="/admin">Admin</NavBar.Link>
        </NavBar>
      )
    ).toBeTruthy();
  });

  test('BackToAdmin component draws correctly', () => {
    const wrapper = shallow(<BackToAdmin />);

    expect(
      wrapper.containsMatchingElement(
        <Card>
          <Button.Light>Back to admin</Button.Light>
        </Card>
      )
    ).toBeTruthy();
  });

  test('BackToAdmin component button takes user to admin page', (done) => {
    const wrapper = shallow(<BackToAdmin />);

    wrapper.find(Button.Light).simulate('click');

    setTimeout(() => {
      expect(location.hash).toEqual('#/admin/');
      done();
    });
  });

  test('Contact component draws correctly', () => {
    const wrapper = shallow(<Contact />);

    expect(
      wrapper.containsMatchingElement(
        <Card title="Contact">
          <ul>
            <li>+47 123 45 678</li>
            <li>fornavn@eposttjeneste.no</li>
            <li>Bosted 12, 3456 Sted</li>
          </ul>
        </Card>
      )
    ).toBeTruthy();
  });

  test('Footer component draws correctly', () => {
    const wrapper = shallow(<Footer />);

    expect(
      wrapper.containsMatchingElement(
        <Card textRight>Laget av Kevin, Hans Petter, Henrik, Bjarne og Mathias</Card>
      )
    ).toBeTruthy();
  });

  test('AdminPage component draws correctly', () => {
    const wrapper = shallow(<AdminPage />);

    expect(
      wrapper.containsMatchingElement(
        <>
          <TitleCard>Mancat admin</TitleCard>
          <ManagementMenu />
          <hr />
          <ProjectListingManagement />
        </>
      )
    ).toBeTruthy();
  });
});
