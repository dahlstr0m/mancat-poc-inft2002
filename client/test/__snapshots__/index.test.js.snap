// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main components tests BackToAdmin component draws correctly 1`] = `
<Card>
  <ButtonLight
    onClick={[Function]}
  >
    Back to admin
  </ButtonLight>
</Card>
`;
